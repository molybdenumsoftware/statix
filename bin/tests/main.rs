mod util {
    #[macro_export]
    macro_rules! test_lint {
        ($tname:ident, $($tail:tt)*) => {
            test_lint!($tname);
            test_lint!($($tail)*);
        };
        ($tname:ident, $($tail:tt)*) => {
                test_lint!($tname);
                test_lint!($($tail)*);
        };
        ($tname:ident) => {
            paste::paste! {
                #[test]
                fn [<$tname _lint>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_lint"), file_path, &["check"]);
                }

                #[test]
                fn [<$tname _fix>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_fix"), file_path, &["fix", "--dry-run"]);
                }
            }

        };
    }
}

fn test_cli(test_name: &str, file_path: &str, args: &[&str]) {
    let output = std::process::Command::new("cargo")
        .arg("run")
        .arg("--")
        .args(args)
        .arg(file_path)
        .output()
        .expect("command runs successfully");

    let stdout = strip_ansi_escapes::strip(output.stdout).unwrap();
    let stdout = String::from_utf8(stdout).unwrap();

    insta::assert_snapshot!(test_name, &stdout);
}

test_lint! {
    collapsible_let_in,
    eta_reduction,
    useless_parens,
    empty_pattern,
    redundant_pattern_bind,
    unquoted_uri,
    empty_inherit,
    deprecated_to_path,
    useless_has_attr,
    repeated_keys,
    empty_list_concat
}
