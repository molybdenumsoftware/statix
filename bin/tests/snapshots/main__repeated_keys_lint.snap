---
source: bin/tests/main.rs
expression: "&stdout"
---
[W20] Warning: Avoid repeated keys in attribute sets
    ╭─[tests/data/repeated_keys.nix:16:5]
    │
 16 │     foo.bar = 1;
    ·     ───┬───  
    ·        ╰───── The key foo is first assigned here ...
 17 │     foo.bar."hello" = 1;
    ·     ───────┬───────  
    ·            ╰───────── ... repeated here ...
 18 │     foo.again = 1;
    ·     ────┬────  
    ·         ╰────── ... and here. Try foo = { bar=...; bar."hello"=...; again=...; } instead.
────╯
[W20] Warning: Avoid repeated keys in attribute sets
    ╭─[tests/data/repeated_keys.nix:23:5]
    │
 23 │     foo.baz.bar1 = 1;
    ·     ──────┬─────  
    ·           ╰─────── The key foo is first assigned here ...
 24 │     foo.baz.bar2 = 2;
    ·     ──────┬─────  
    ·           ╰─────── ... repeated here ...
 25 │     foo.baz.bar3 = 3;
    ·     ──────┬─────  
    ·           ╰─────── ... and here (2 occurrences omitted). Try foo = { baz.bar1=...; baz.bar2=...; baz.bar3=...; } instead.
────╯
